### DISCUSSION TITLE
How to skip query string percent encoding?

### DISCUSSION THREAD

**Discussion by trim21** [encode/httpx#2844] [Q&A] [answered]
I have a url looks like `"https://example.com/?TQC/KWC0qaYyEg=="`

it's a base64 encoded string so it contains `/`.

`requests` just use the raw url input, and I find that httpx will try to encode `/` with `%2F`, so it's not the expected url, and I can't find how to disable this encoding.

```text
import httpx


raw = "https://example.com/?TQC/KWC0qaYyEg=="
url = httpx.URL(raw)

print(url)  # https://example.com/?TQC%2FKWC0qaYyEg==
print(httpx.get(raw).request.url)  # https://example.com/?TQC%2FKWC0qaYyEg==
```

**Comment by lovelydinosaur** [comment]
I can provide some pointers here if you want to dig into this.

First up... which existing issues/discussions do we currently have in the project around this?

**Reply by trim21** [reply]
reason is a pr merged in 0.25.0: Always encode forward slashes as `%2F` in query parameters (https://github.com/encode/httpx/pull/2723)

I want to find a way to bypass this...

**Reply by lovelydinosaur** [reply]
Useful avenues...

* Are would you able to do this with `httpcore`?
* Which encoding behaviour does `curl` default to here?
* Why does your webserver need to differentiate between the encoded / unencoded variants here? *(Aka... "I want to find a way to bypass this" ... why?)*

**Reply by trim21** [reply]
Yes, this can be done with `httpcore` by `httpcore.request("GET", "https://example.com/?TQC/KWC0qaYyEg==")`

curl also send `/` unescaped.

```text
> GET /?TQC/KWC0qaYyEg== HTTP/2
> Host: example.com
> user-agent: curl/7.74.0
> accept: */*
```

because this webserver read whole querystring as base64 input, it didn't parse base64 as key/value pairs. and I have no control how this webserver works, I'm only user.

**Reply by lovelydinosaur** [reply]
> Yes, this can be done with `httpcore` by `httpcore.request("GET", "https://example.com/?TQC/KWC0qaYyEg==")`

Where are we passing the URL from `httpx` to `httpcore`?

*(I hope this isn't too tedious... I'm just working through the same process I would use when figuring out a resolution here)*

**Reply by trim21** [reply]
https://github.com/encode/httpx/blob/733595037ac26d272f74ee6d702150b017913d11/httpx/_transports/default.py#L214-L227

**Reply by lovelydinosaur** [reply]
> requests just use the raw url input

Curiously this is at odds with [the conversation here](https://github.com/encode/httpx/discussions/2661#discussioncomment-5659997) which indicated that requests *does* URL encode the '/' character.

I *think* that comment was imprecise, and the *actual* behaviour is that `requests` will URL encode the `/` character when it's used as a value in `params={...}` but will leave it alone when passed as part of a URL string.
 
(?)

---

Anyways, let's just knock one thing off here to sharpen things up a little... `url.raw_path` would be better named as `url.target`. Because it's the raw byte string that's used as the `target` part of a standard non-proxied HTTP/1.1 request. That would be a more precise clear name.

Let's pretend that is for the below examples, just because I think it helps make things more clear.

At this point we can *isolate the conversation* down to one around the `httpx.URL()` API...

```python
url = httpx.URL("https://example.com/?abc/ 123")
url.target  # What is the expected behaviour here?
```

```python
url = httpx.URL("https://example.com/", params={"key": "abc/ 123"})
url.target  # What is the expected behaviour here?
```

```python
url = httpx.URL("https://example.com/?abc%2F%20123")
url.target  # What is the expected behaviour here?
```

```python
url = httpx.URL("https://example.com/", params={"key": "abc%2F%20123"})
url.target  # What is the expected behaviour here?
```

**Reply by trim21** [reply]
it doesn't always encode `/`
```python
import requests


r = requests.request("GET", "https://example.com/?TQC/KWC0qaYyEg==")
print(r.request.url)  # https://example.com/?TQC/KWC0qaYyEg==
```

**Reply by lovelydinosaur** [reply]
Indeed...

Does not percent encode the `"/"` character in a URL string.

```python
>>> r = requests.request("GET", "https://example.com/?TQC/KWC0qaYyEg==")
>>> r.request.url
'https://example.com/?TQC/KWC0qaYyEg=='
```

Percent encodes the `"/"` character in a param value.

```python
>>> r = requests.request("GET", "https://example.com/", params={"base64": "TQC/KWC0qaYyEg=="})
>>> r.request.url
'https://example.com/?base64=TQC%2FKWC0qaYyEg%3D%3D'
```

Percent encodes the `" "` character in a URL string.

```python
>>> r = requests.request("GET", "https://example.com/?url with spaces ")
>>> r.request.url
'https://example.com/?url%20with%20spaces%20'
```

Plus encodes the `" "` character in a param value.

```python
>>> r = requests.request("GET", "https://example.com/", params={"key": "url with spaces"})
>>> r.request.url
'https://example.com/?key=url+with+spaces'
```

**Reply by trim21** [reply]
so what's your opinion on this issue? `yarl` (`aiohttop`'s url lib) have a argument `encoded` to skip encoding url again.

**Reply by meiers** [reply]
Dear maintainers,
I stumbled over something similar today because one of my automated tests failed. Here it seems `%` symbols are unintentionally encoded.

Bug?

```python
for q in [
    "__%2F__%23__%3F__%26__%0A__",
    "__/__%23__%3F__%26__%0A__",
    "__/__%23__?__%26__%0A__",
    # "__/__%23__?__%26__\n__", this is supposed to fail
]:
    response = test_client.get(f"/samples?sample_id_or_name={q}")
    assert response.status_code == 200
    print(f"search sample {response.request.url}: {response.status_code}")
>   assert response.status_code == 200
E    assert 404 == 200
E     +  where 404 = <Response [404 Not Found]>.status_code
```
```
search sample http://testserver/samples?sample_id_or_name=__%2F__%23__%3F__%26__%0A__: 200
search sample http://testserver/samples?sample_id_or_name=__%2F__%2523__%253F__%2526__%250A__: 404
```
The second line has all % characters except the first one encoded as %25.

This is using the `TestClient` from fastapi. These are the installed versions:

```
fastapi==0.99.1
httpx==0.25.0
pytest==7.4.2
requests==2.31.0
starlette==0.27.0
urllib3==1.26.16
```

After pinning `httpx<0.25` the test passes again (but also `urllib3` changed, not sure if coincidently):
```
fastapi==0.99.1
httpx==0.24.1
pytest==7.4.2
requests==2.31.0
starlette==0.27.0
urllib3==2.0.4
```

**Reply by lloydzhou** [reply]
> versions: 0.6.7, 0.6.8, 0.7.0, 0.7.1, 0.7.2, 0.7.3, 0.7.4, 0.7.5, 0.7.6, 0.7.7, 0.7.8, 0.8.0, 0.9.0, 0.9.1, 0.9.2, 0.9.3, 0.9.4, 0.9.5, 0.10.0, 0.10.1, 0.11.0, 0.11.1, 0.12.0, 0.12.1, 0.13.dev0, 0.13.0.dev1, 0.13.0.dev2, 0.13.0, 0.13.1, 0.13.2, 0.13.3, 0.14.0, 0.14.1, 0.14.2, 0.14.3, 0.15.0, 0.15.1, 0.15.2, 0.15.3, 0.15.4, 0.15.5, 0.16.0, 0.16.1, 0.17.0, 0.17.1, 0.18.0, 0.18.1, 0.18.2, 0.19.0, 0.20.0, 0.21.0, 0.21.1, 0.21.2, 0.21.3, 0.22.0, 0.23.0, 0.23.1, 0.23.2, 0.23.3, 0.24.0, 0.24.1, 0.25.0, 0.25.1, 1.0.0b0

same error in `httpx==0.25.0` and `httpx==0.25.1`

test passes in `httpx==0.24.1` and `httpx==1.0.0b0`

**Reply by lloydzhou** [reply]
in `urllib.parse.urlparse` keep the existing escape string
but `httpx._urlparse.urlparse` encoding the existing escape string


```python
>>> from urllib.parse import urlparse
>>> from httpx._urlparse import urlparse as hurlparse
>>> url = 'https://example.com?sig=iGyXM1mSehS4ONN5%2B2GillQ%2BYawyh5EryToDT/Bzf90%3D'
>>> urlparse(url)
ParseResult(scheme='https', netloc='example.com', path='', params='', query='sig=iGyXM1mSehS4ONN5%2B2GillQ%2BYawyh5EryToDT/Bzf90%3D', fragment='')
>>> hurlparse(url)
ParseResult(scheme='https', userinfo='', host='example.com', port=None, path='', query='sig=iGyXM1mSehS4ONN5%252B2GillQ%252BYawyh5EryToDT%2FBzf90%253D', fragment=None)
>>>
```

**Reply by dhensen** [reply]
I'm downgrading. This has been driving me crazy and losing my sanity for the last couple of days. Why? Because working on multiple projects with different versions of httpx and also other http libraries. At least now we know there is a discrepancy in httpx that needs to be solved. Let me know if you need anything
```
poetry add "httpx<0.25"

Updating dependencies
Resolving dependencies... (0.6s)

Package operations: 0 installs, 2 updates, 0 removals

  • Downgrading httpcore (1.0.1 -> 0.17.3)
  • Downgrading httpx (0.25.1 -> 0.24.1)
```

**Reply by alexmaurizio** [reply]
Upping the thread here too, because we had to downgrade our entire stack to 0.24.1 from 0.25.x due to this change.
This change does break a number of external and internal APIs and there is not a sane way to work around this "feature" without locking to previous version.

**Reply by lovelydinosaur** [reply, marked as answer]
Resolved in https://github.com/encode/httpx/pull/2990
